### PR TITLE
[sw, host] Add initial ROM_EXT image signer tool infrastructure

### PR DESCRIPTION
**Pull request by silvestrst**
# Current state

This PR pulls the required dependencies and produces an `elf` file that prints "Hello World!".

- The actual signer implementation I think can be done in parallel with minimal pain.

- The `objcopy` of the image blob can be also done more or less in parallel.

# Background

ROM_EXT images signing tool is required. The idea is simple - given an image with "empty" manifest, the tool must populate and signe the image.

We have decided to use `rust` and `cargo` to write the signer in, which must use the `ring` crate for actual signing.

# This pull-request consists of three changes

I will aim to describe in the sections below the rationale and questions that I hope you could help me find the answers for .

## [util, cargo] Rename `build_tock.sh` into `invoke_cargo.sh`

This is the script that has been written by @jon-flatley for the downstream Tock invocation. It aims to address some of `meson` shortcomings.

I thought it would be a good idea to reuse this script for the purpose of this PR, until (if) we find a better way of addressing this issues.

## [sw, host] Add initial ROM_EXT image signer tool infrastructure

The main change in the series, most `meson.build` functionality has been taken from the `sw/device/tock/meson.build`.

### Questions

- [x] Is `sw/host/rom_ext_image_signer` an appropriate directory for this tool?
Renamed as suggested in the review comments to `sw/host/rom_ext_image_tools`, but the `host` is an appropriate top level directory. 

- [x] Do we have any restrictions on what dependencies we can use?
So far `ring` and `hjson-serde`, the rest (none at this point) need to be approved.

- [ ] Is this tool supposed to be built with every invocation of `ninja`, or should be have an option to add/remove this from the build through `meson_init`?

- [ ] This partially touches the next change, but do we want to **always** make the tool use local dependencies, which we can easily vendor in with the script? From the user perspective this is similar to fetching the python requirements.

- [x] Could you please help me address the `TODOs` sprinkled in the code, for example, is it enough to just use a default target (see `meson.build`)?
I think we have addressed all/most inline TODOs.

## [sw, host] Add ROM_EXT signer `Cargo` dependencies vendoring

This change introduces a bash script to vendor in dependencies. I deliberately didn't go out of my way to make it pretty in case you guys have strong opinions on how this should be implemented.

At this point however, it is more important to answer larger questions, summarised in the section below.

### Questions

- [x] Is `cargo vendor` a sufficient way to gather dependencies?
Quite likely that it will come in useful later on, at the moment we should just fetch dependencies online for simplicity.

- [ ] As per questions from the previous change, do we want to always build the tool from the local dependencies, or add this on-demand via the script?

### CODE REVIEW

File: sw/host/rom_ext_image_signer/Cargo.toml
Comment (lenary) [inline sw/host/rom_ext_image_signer/Cargo.toml:15]: panic = abort isn't correct for host builds where we know we have an OS, and in dev we probably don't want opt-level = `z` at all on hosts (this is required on embedded to keep image size down).

File: sw/host/rom_ext_image_signer/vendor-dependencies.sh
Comment (lenary) [inline sw/host/rom_ext_image_signer/vendor-dependencies.sh:20]: We want `--locked`, so we always pull the versions specified in Cargo.lock, rather than any versions that satisfy the constraints in Cargo.toml.

File: sw/host/rom_ext_image_signer/src/main.rs
Comment (lenary) [inline sw/host/rom_ext_image_signer/src/main.rs:1]: Nit: Licence Header :)

**Review by lenary — COMMENTED** [review]
Answers to your questions on second commit:
- `sw/host/rom_ext_image_tools` might be a better name, just in case we get others - we may have one cargo file which creates multiple binaries, potentially.
- You need to run potential dependencies past Garret (and me, but mainly garret). We have already approved `ring`, as noted.
- At this stage, I think we want to mainly build from online dependencies, to keep things easier. Thanks for the work on vendoring, that is going to be hugely useful when we want to create builds in more restricted environments.
- Hopefully a bunch of TODOs are answered inline.

Answers to questions on third commit:
- `cargo vendor` seems built in, and looks like it solves the problem, so I'm happy with it.

File: sw/host/rom_ext_image_signer/meson.build
Comment (jon-flatley) [reply sw/host/rom_ext_image_signer/meson.build:6]: I agree, we should make sure that all `build_by_default` targets don't depend on `required: false` tools.

File: sw/host/rom_ext_image_signer/rust-toolchain
Comment (jon-flatley) [reply sw/host/rom_ext_image_signer/rust-toolchain:1]: We should avoid nightly toolchains unless we have a good reason to use them. FWIW I think if we don't need to override the toolchain and aren't passing any `RUSTFLAGS` we don't actually need to use a wrapper script, we can just call cargo directly. I still think it's a better idea to have a wrapper like I talked about above.

File: sw/host/rom_ext_image_signer/Cargo.lock
Comment (mcy) [inline sw/host/rom_ext_image_signer/Cargo.lock:1]: I don't believe Cargo.lock files are usually checked in. Are we doing this on purpose, or should we fix the gitignore?

File: sw/host/rom_ext_image_signer/Cargo.toml
Comment (mcy) [reply sw/host/rom_ext_image_signer/Cargo.toml:24]: I strongly suggest pinning to an exact version, because it avoids having to trust upstreams from doing the right thing (I don't have very much faith in the ecosystem not making mistakes).

File: sw/host/rom_ext_image_signer/meson.build
Comment (mcy) [reply sw/host/rom_ext_image_signer/meson.build:8]: > In fact, after reading up some more, we shouldn't be specifying --target at all for host builds, which probably makes invoke_cargo.sh a little more complex, but still simple enough to share between host and device builds.

Just to +1, this is correct. I think the right thing to do is to make invoke_cargo.sh not emit a --target command when the relevant environment variable is missing, which can be done like this:
```
"${TARGET:+--target=}${TARGET}"
```
`${var:+foo}` will emit `foo` if `var` is set.

File: sw/host/rom_ext_image_signer/meson.build
Comment (mcy) [reply sw/host/rom_ext_image_signer/meson.build:49]: Do we build in release or debug consistently? If so, the cargo output is easy to find. (E.g. target/release/my_binary). Note that this is only true if you don't pass --target.

File: sw/host/rom_ext_image_signer/rust-toolchain
Comment (mcy) [reply sw/host/rom_ext_image_signer/rust-toolchain:1]: ++ Avoid nightly like the plague unless you desperately need asm!

File: .gitignore
Comment (mcy) [inline .gitignore:57]: You dropped the trailing newline. =)

File: sw/host/rom_ext_image_signer/vendor-dependencies.sh
Comment (mcy) [inline sw/host/rom_ext_image_signer/vendor-dependencies.sh:1]: It's not immediately clear what this script is doing. Would you mind adding a comment explaining it?

File: sw/host/rom_ext_image_signer/vendor-dependencies.sh
Comment (mcy) [inline sw/host/rom_ext_image_signer/vendor-dependencies.sh:45]: Something tells me this script is complex enough it aught to be a python script...

File: sw/host/rom_ext_image_signer/Cargo.lock
Comment (lenary) [reply sw/host/rom_ext_image_signer/Cargo.lock:1]: From [Cargo.toml vs Cargo.lock (docs.rust-lang.org)](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html)
> If you’re building an end product, which are executable like command-line tool or an application, or a system library with crate-type of staticlib or cdylib, check Cargo.lock into git.

File: sw/host/rom_ext_image_signer/Cargo.toml
Comment (alphan) [reply sw/host/rom_ext_image_signer/Cargo.toml:24]: +1 to pinning to an exact version to avoid surprises.

**Review by silvestrst — COMMENTED** [review]
**Response to @mcy comments (for some reason cannot respond directly to some):**
> You dropped the trailing newline. =)

Fixed.

> It's not immediately clear what this script is doing. Would you mind adding a comment explaining it?
> Something tells me this script is complex enough it aught to be a python script...

When we get to vendoring, this is just an example for now.

> ++ Avoid nightly like the plague unless you desperately need asm!

Dropped the toolchain file.

> Could you make a list of them?

TODO

**Response to @lenary comments:**

> sw/host/rom_ext_image_tools might be a better name, just in case we get others - we may have one cargo file which creates multiple binaries, potentially.

Done.

> You need to run potential dependencies past Garret (and me, but mainly garret). We have already approved ring, as noted.

Noted, so far ring and hjson-serde seems non-contraversial.

> At this stage, I think we want to mainly build from online dependencies, to keep things easier.

Agreed.

>Thanks for the work on vendoring, that is going to be hugely useful when we want to create builds in more restricted environments.

No worries, hopefully that will come in handy at some point.

> Hopefully a bunch of TODOs are answered inline.

Thanks!

**Comment by silvestrst** [comment]
I think might be a bad rebase, will update shortly.

**Comment by silvestrst** [comment]
@lenary you were not sure whether meson handles the string of flags correctly. I have added `set x` to the script to check, and that is how the commenad is invoked from within the `invoke_cargo.sh`:
```
/usr/bin/cargo + build --manifest-path=/home/svt/development/repos/opentitan/sw/host/rom_ext_image_tools/Cargo.toml --target-dir=/home/svt/development/repos/opentitan/build-out/sw/host/rom_ext_image_tools --release
```
So that seems to be fine.

Another concern you had was regarding the empty toolchain file, although it works (cargo knows how to deal with a `+` without any toolchain specified), I think I will change it to not emit a `+` for an empty toolchain file. Also, as the toolchain file is in `depend_files`, even if we don't need a specific toolchain, we still should have an empty file.

I am about to push the updates.

**Comment by silvestrst** [comment]
Ping

File: sw/host/rom_ext_image_tools/meson.build
Comment (lenary) [inline sw/host/rom_ext_image_tools/meson.build:30]: For Follow-up: I'm worried by these two - and how this is dealt with in `invoke_cargo.sh`. I think we can find a better way, but that should not block this PR

File: sw/host/rom_ext_image_tools/meson.build
Comment (lenary) [inline sw/host/rom_ext_image_tools/meson.build:62]: Nit: We don't call things `.elf` if they're host binaries - they just get no extension.
```suggestion
  output: 'rom_ext_signer',
```

**Review by lenary — APPROVED** [review]
LGTM. Two issues, both can be solved *after* landing tbh.

File: .gitignore
Comment (gkelly) [inline .gitignore:60]: Is there any way these can be made to live under `third_party/sw/vendored_dependencies`? That would fit with what our best-practices are for project structure.

File: sw/host/rom_ext_image_signer/vendor-dependencies.sh
Comment (silvestrst) [reply sw/host/rom_ext_image_signer/vendor-dependencies.sh]: This is something for later (just an example at the moment)

File: .gitignore
Comment (silvestrst) [reply .gitignore:60]: Opened a follow-up issue: https://github.com/lowRISC/opentitan/issues/4351

**Comment by silvestrst** [comment]
Unresolved comments are left for the follow-ups, see: https://github.com/lowRISC/opentitan/issues/4351

**Pull request merged**
